elkodon/src/raw_sample.rs

```diff
@@ -0,0 +1,193 @@
+use crate::message::Message;
+
+use core::fmt;
+
+/// A `*const Message<Header, Data>` non-zero sample pointer to the message.
+#[repr(transparent)]
+pub(crate) struct RawSample<Header, Data> {
+    message: *const Message<Header, Data>,
+}
+
+impl<Header, Data> RawSample<Header, Data> {
+    /// Creates a new `RawSample`.
+    ///
+    /// # Safety
+    ///
+    /// `message` must be non-null.
+    #[inline]
+    pub(crate) unsafe fn new_unchecked(message: *const Message<Header, Data>) -> Self {
+        debug_assert!(
+            !message.is_null(),
+            "RawSample::new_unchecked requires that the message pointer is non-null"
+        );
+        Self { message }
+    }
+
+    /// Creates a new `RawSample`.
+    #[allow(dead_code)]
+    #[inline]
+    pub(crate) fn new(message: *const Message<Header, Data>) -> Option<Self> {
+        if !message.is_null() {
+            // SAFETY: `message` pointer is checked to be non-null
+            Some(unsafe { Self::new_unchecked(message) })
+        } else {
+            None
+        }
+    }
+
+    /// Acquires the underlying message as `*const` pointer.
+    #[must_use]
+    #[inline(always)]
+    pub(crate) fn as_ptr(self) -> *const Message<Header, Data> {
+        self.message
+    }
+
+    /// Acquires the underlying message as reference.
+    #[must_use]
+    #[inline(always)]
+    pub(crate) fn as_ref(&self) -> &Message<Header, Data> {
+        // SAFETY: `self.as_ptr()` returns a non-null ptr and `Data` is either the actual message type or wrapped by a `MaybeUninit` which makes a reference to `Message::data` safe
+        unsafe { &(*self.as_ptr()) }
+    }
+
+    /// Acquires the underlying header as reference.
+    #[must_use]
+    #[inline(always)]
+    pub(crate) fn as_header_ref(&self) -> &Header {
+        &self.as_ref().header
+    }
+
+    /// Acquires the underlying data as reference.
+    #[must_use]
+    #[inline(always)]
+    pub(crate) fn as_data_ref(&self) -> &Data {
+        &self.as_ref().data
+    }
+}
+
+impl<Header, Data> Clone for RawSample<Header, Data> {
+    #[inline(always)]
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Header, Data> Copy for RawSample<Header, Data> {}
```

Comment:
Would it be possible to annotate the `RawSample` with `#[derive(Copy)]`?

Reply:
It must be `#[derive(Copy, Clone)]`. I'm not sure why I did the manual impl in iceoryx-rs. Maybe the derive does not work with raw pointers. Have to check this again.

Reply:
Unfortunately this does not work due to the raw pointer. It has to be implemented manually.